pkg/storage/tsdb/multi_file_index.go

```diff
@@ -92,80 +92,68 @@ func (i *MultiIndex) forIndices(ctx context.Context, from, through model.Time, f
 	return results, nil
 }
 
-func (i *MultiIndex) GetChunkRefs(ctx context.Context, userID string, from, through model.Time, shard *index.ShardAnnotation, matchers ...*labels.Matcher) ([]ChunkRef, error) {
+func (i *MultiIndex) GetChunkRefs(ctx context.Context, userID string, from, through model.Time, res *[]ChunkRef, shard *index.ShardAnnotation, matchers ...*labels.Matcher) error {
+	*res = (*res)[:0]
 	groups, err := i.forIndices(ctx, from, through, func(ctx context.Context, idx Index) (interface{}, error) {
-		return idx.GetChunkRefs(ctx, userID, from, through, shard, matchers...)
+		refs := ChunkRefsPool.Get()
+		err := idx.GetChunkRefs(ctx, userID, from, through, &refs, shard, matchers...)
```

Comment:
if the final append was inside here you wouldn't need to Get from the pool multiple times, just reset the slice for each indice.

Same for the Multi Series.

Reply:
I used independent slices because `i.forIndices` runs all these calls in parallel.